### PR TITLE
API: Preprocess assigns wrong query bindings when there is more than one duplicated value

### PR DESCRIPTION
**Pull request by joselcvarela**
## Scope
When using [Aggregation or Grouping](https://docs.directus.io/reference/query.html#aggregation-grouping) together with the `search` query we were getting the error:
```
{"errors":[{"message":"An unexpected error occurred.","extensions":{"code":"INTERNAL_SERVER_ERROR"}}]}
```

And on server logs we saw:
```
could not determine data type of parameter $5
```

After some investigation we found this to be happening when database query bindings had more than one duplicated value. For example, binding values looked like this:
```js
[ "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "%search term%", "published", "published", "draft", 100, ]
```

This caused the SQL query to look like something like
```
select ... from ... where ... = $1 ... = $2 ... =$33
```

The `$33` should be something like `$3` and that's what this PR tries to fix.

The underlying problem was that when looking up the index to use for a binding that was encountered before (`matchIndex !== bindingIndices[matchIndex]`) instead of using the binding index of the new, deduplicated binding, the index of it's original first occurrence tracked by `bindingsIndices[matchIndex]` is used. 

The fix is to replace the index of the first occurrence in the `bindingsIndices` with the newly assigned index when it is first encountered in the replacement loop. When it is encountered again we have to do a double lookup, where the first `bindingIndices[matchIndex]` yields the index of the first occurrence, and resolving that yields the newly assigned index for that binding value.

What's changed:

- Fixed the binding preprocessing to account for multiple occurrences 
- Added a test that catches the incorrect behavior on main

## Potential Risks / Drawbacks

- None

## Review Notes / Questions

- None

### CODE REVIEW

File: api/src/database/helpers/schema/utils/preprocess-bindings.ts [hunk 1]
```diff
@@ -1,46 +1,72 @@
...
-		if (prevIndex !== -1) {
-			bindingIndices[i] = prevIndex;
-		} else {
+		if (prevIndex === -1) {
+			// The first time the value is encountered, set the index lookup to the current index
 			bindingIndices[i] = i;
+		} else {
+			// The value has been encountered before, set the index lookup to the first occurrence index
+			bindingIndices[i] = prevIndex;
 		}
```
Comment (hanneskuettner) [inline api/src/database/helpers/schema/utils/preprocess-bindings.ts:31]: This is just switching the order of the if statement, as it is more intuitive to read if the first block is the first time a value is encountered (`prevIndex === -1`)

**Comment by hanneskuettner** [comment]
The reason this worked in the test / review previously is that the there either weren't more than one duplicated binding, or the duplicated bindings showed up with their first occurrences being in sequence, so `bindingIndices[matchIndex]` matched up with the newly assigned index, eg. 0, 1, 2, etc.

**Comment by hanneskuettner** [comment]
@licitdev using a `Map` here definitely is the smarter approach, very nice! I've looked at your changes and they LGTM!

**Review by licitdev — APPROVED** [review]
@hanneskuettner Thanks for the confirmation and improvement! 🫡

**Pull request merged**